core/src/block.rs

```diff
@@ -19,894 +19,630 @@ use iroha_data_model::{block::*, events::prelude::*, transaction::prelude::*};
-            )
-        }
+    /// Check if a block has transactions that are already in the blockchain.
+    fn has_committed_transactions(&self, wsv: &WorldStateView) -> bool {
+        self.transactions
+            .iter()
+            .any(|transaction| transaction.is_in_blockchain(wsv))
+            || self
+                .rejected_transactions
+                .iter()
+                .any(|transaction| transaction.is_in_blockchain(wsv))
     }
+}
 
-    impl CandidateBlock {
-        /// Calculate the hash of the current block.
-
-        #[inline]
-        pub fn hash(&self) -> HashOf<Self> {
-            HashOf::new(&self.header).transmute()
-        }
-
-        /// Return signatures that are verified with the `hash` of this block, removing all other signatures.
-
-        #[inline]
-        pub fn retain_verified_signatures(&mut self) -> impl Iterator<Item = &SignatureOf<Self>> {
-            self.signatures.retain_verified_by_hash(self.hash())
-        }
-
-        /// Check if there are no transactions in this block.
-        #[inline]
-        pub fn is_empty(&self) -> bool {
-            self.transactions.is_empty() && self.rejected_transactions.is_empty()
-        }
-
-        /// Revalidate a block against the current state of the world.
-        ///
-        /// # Errors
-        /// - Block is empty
-        /// - Block has committed transactions
-        /// - There is a mismatch between candidate block height and actual blockchain height
-        /// - There is a mismatch between candidate block previous block hash and actual latest block hash
-        /// - Block header transaction hashes don't match with computed transaction hashes
-        /// - Error during revalidation of individual transactions
-        #[allow(clippy::too_many_lines)]
-        pub fn revalidate<const IS_GENESIS: bool>(
-            self,
-            transaction_validator: &TransactionValidator,
-            wsv: &WorldStateView,
-            latest_block: Option<HashOf<VersionedCommittedBlock>>,
-            block_height: u64,
-        ) -> Result<SignedBlock, eyre::Report> {
-            if self.is_empty() {
-                bail!("Block is empty");
-            }
-
-            if self.has_committed_transactions(wsv) {
-                bail!("Block has committed transactions");
-            }
+impl Revalidate for VersionedCommittedBlock {
+    /// Revalidate a block against the current state of the world.
+    ///
+    /// # Errors
+    /// - Block is empty
+    /// - Block has committed transactions
+    /// - There is a mismatch between candidate block height and actual blockchain height
+    /// - There is a mismatch between candidate block previous block hash and actual latest block hash
+    /// - Block header transaction hashes don't match with computed transaction hashes
+    /// - Error during revalidation of individual transactions
+    #[allow(clippy::too_many_lines)]
+    fn revalidate<const IS_GENESIS: bool>(
```

Comment:
In case of `CommittedBlock` we only want to revalidate number of signatures and check transactions hashes.

Reply:
> There is a bug in validation

in general, I don't we should be structuring our code around an assumption that it's buggy. Maybe this case is different, I won't go into it deeply

Reply:
> in general, I don't think we should be structuring our code around an assumption that it's buggy. Maybe this case is different, I won't go into it deeply

I agree. But I believe this is a case of "don't trust, verify" that we should be upholding as a blockchain project.

Reply:
@appetrosyan Can we have your opinion on this matter?

Reply:
I agree with both to an extent. 

Right now, we should do the full re-verification, until we conduct a full audit of `sumeragi` our main assumption should be that the code is exceedingly buggy, unless proven otherwise. 

In the future, when we can formally verify that the business logic was executed correctly; which at present doesn't seem possible, but given several audits, and thorough test coverage and fuzzing, should be doable; we should short-circuit double checking.